feat(report): add granular CPU/mem configurability

andrewazores opened the pull request:
Fixes #5
Fixes #6

hareetd commented on src/main/docker/Dockerfile.jvm:
```diff
@@ -23,7 +23,7 @@
 FROM registry.access.redhat.com/ubi8/ubi-minimal:8.5
 
-ARG JAVA_PACKAGE=java-11-openjdk-headless
+ARG JAVA_PACKAGE=java-17-openjdk-headless
```
Was it updating the Java package that fixed the HotSpot bug?

andrewazores replied:
Seems like it.

hareetd reviewed (commented):
I packaged this branch and ran a smoketest instance of https://github.com/cryostatio/cryostat/pull/779, report generation worked using the ~10MB file Janelle linked (https://github.com/cryostatio/cryostat/pull/779#discussion_r766944173) that was previously failing for her.

hareetd commented:
Any chance you could link the ~150MB file you mentioned in the corresponding issue #5, perhaps through Google Drive? I'd like to test out the PR using that as well.

andrewazores commented:
> Any chance you could link the ~150MB file you mentioned in the corresponding issue #5, perhaps through Google Drive? I'd like to test out the PR using that as well.

I'll email you a link to it.

hareetd commented:
> If I upload the recording, then click View Report, it hangs.

> If I upload the recording, then click the dropdown, the report shows up nicely in the dropdown after 5s, and a subsequent click > on View Reports show the report in a new tab.

I spoke too soon, Janelle's 10MB one and the ~150MB one you sent me are still exhibiting the above behaviour. I originally just tested the dropdown report for the 10MB recording and not View Report in a new tab. 

As for the ~480MB file, both the View Report and dropdown don't work, unfortunately. 

My workflow was to upload the files to archives and for each one do: View Report --> Dropdown --> View Report. I did this in order of smallest to largest. The relevant logs are linked below.

[Cryostat Logs](https://gist.github.com/hareetd/81b204926a68b359acdcefeb41099c94)

[Cryostat Reports Container Logs](https://gist.github.com/hareetd/644a889ed8169a3ccccdadcab8a00c0b)

andrewazores commented:
The `Connection was closed` one looks like it's failing because the `POST` sending the data from `cryostat` to `-reports` ended up timing out, I think? Maybe we want configuration parameters to control the maximum time that Cryostat will wait for a response and the maximum time that -reports will spend on reading a request's POST body (if it does have such a time limit currently). Having a timeout at some point does seem desirable because we don't want to have Cryostat API requests that just hang indefinitely if the -reports container(s) are very slow or the network link between them is very slow.

> I spoke too soon, Janelle's 10MB one and the ~150MB one you sent me are still exhibiting the above behaviour. I originally just tested the dropdown report for the 10MB recording and not View Report in a new tab.

I've been trying to figure out where all the memory usage is going because it does seem like the sidecar -reports container should be able to handle larger recording files than it currently does, but I haven't had too much success yet. That cryostatio/cryostat#779 PR is set up so that -reports has JDP and JMX enabled, so you can interact with it as any other target and use Cryostat to collect a Profiling JFR recording from the reports sidecar.

You can try increasing the CPU and memory allocation given to the container by editing those parameters in the `smoketest.sh` flags passed to `podman` and set in the `JAVA_OPTIONS`. The default 1 CPU/512MB doesn't go as far as I'd like it to, that's for sure.

You can also use the `run.sh` from this PR to test the capabilities with different resource allocations more quickly and easily, ex. `CPUS=2 MEMORY=2G sh run.sh`, then`http :8080/report file@my.jfr`.

andrewazores commented:
I tried opening the 150MB .jfr file using JMC, while also running a Profiling recording on JMC itself, and found that the heap memory usage there is also dominated by allocations that stem from the report generation. It ends up looking very similar to the -reports allocations and heap usage. So, I think the 150MB file or any larger ones just really do take quite a lot of memory to process. Opening it in JMC normally makes it seem pretty quick and easy but that's because JMC as a desktop application is able to use all of your workstation's CPU cores and can allocate a lot of heap space (probably about 1/4 of your workstation's RAM) by default, and these resource allocations completely dwarf the limits we're testing -reports with.

andrewazores commented:
```bash
$ cd workspace/jmc
$ $EDITOR ./target/products/org.openjdk.jmc/linux/gtk/x86_64/jmc.ini
# add "-Xms512M", "-Xmx512M", "-XX:ActiveProcessorCount=1" to the flags
$ ./target/products/org.openjdk.jmc/linux/gtk/x86_64/jmc
```

I tried this out to launch the JMC desktop application with similar constraints that we apply to cryostat-reports (512M memory, 1 CPU core), then tried opening the 150MB .jfr with it. JMC also failed to open it, although JMC's failure handling is nice and graceful:

![Screenshot_2021-12-22_14-12-02](https://user-images.githubusercontent.com/3787464/147143467-e8d5f4db-8ff0-421a-9b65-2abe7069b9b8.png)

It seems to recognize that the recording is too large, and then presents a range selector to allow you to choose a narrower time window to try to analyze instead. After a couple of trials it seems the width of the bar corresponds to how much memory is currently available to JMC, so there appears to be some heuristic that JMC applies here to estimate how large of a recording it can handle given its memory constraints.

If there is such a utility in the JMC code somewhere doing this approximation then it would be very neat if we could reuse that and expose it in the cryostat-core report generator. Then we can reject requests for recordings that we estimate to be too large, rather than wasting resources trying to process the recording and ending in an OOM or TimeoutException.

hareetd commented on run.sh:
```diff
@@ -0,0 +1,29 @@
+    --cpus "${CPUS}" \
+    --memory "${MEMORY}" \
+    --publish 8080:8080 \
+    --env JAVA_OPTIONS="-XX:ActiveProcessorCount=${CPUS} -XX:+Use${GC}GC -Dorg.openjdk.jmc.flightrecorder.parser.singlethreaded=${SINGLETHREAD_JFR_PARSE}" \
```
It seems like `--memory "${MEMORY}"` doesn't actually set the maximum available memory for the containerized JVM because on the output ` Runtime.getRuntime().maxMemory() / (1024 * 1024))` from `ReportResource` when starting `run.sh`. However, adding the `-Xmx${MEMORY}` flag to `JAVA_OPTIONS` looks like it does the trick.

andrewazores replied:
Huh. Well, `-Xmx` sets the maximum allocation pool size, which is only a portion of the total memory that the JVM will allocate. The podman `--memory` flag should set the total memory limit for the container, not just the allocation pool size.

If the JVM isn't properly detecting its container memory limits then it would be choosing bad ratios for ex. heap size vs other memory pools, which could lead to weird OOM and performance issues when we're running workloads like this.

Later JDK 11s and especially the more recent JDK 17 are supposed to have proper container ergonomics and support for limits like this however.

All that said:

```
$ MEMORY=512M sh run.sh 
+ set -e
+ '[' -z '' ']'
+ CPUS=1
+ '[' 1 -gt 1 ']'
+ GC=Serial
+ SINGLETHREAD_JFR_PARSE=true
+ '[' -z 512M ']'
+ podman run --user 0 --cpus 1 --memory 512M --publish 8080:8080 --env 'JAVA_OPTIONS=-XX:ActiveProcessorCount=1 -XX:+UseSerialGC -Dorg.openjdk.jmc.flightrecorder.parser.singlethreaded=true' --rm -it quay.io/cryostat/cryostat-reports:latest
exec java -XX:ActiveProcessorCount=1 -XX:+UseSerialGC -Dorg.openjdk.jmc.flightrecorder.parser.singlethreaded=true -cp . -jar /deployments/quarkus-run.jar
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/deployments/lib/main/org.jboss.slf4j.slf4j-jboss-logmanager-1.1.0.Final.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/deployments/lib/main/org.slf4j.slf4j-jdk14-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Slf4jLoggerFactory]
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2021-12-22 19:43:34,890 INFO  [io.cry.rep.ReportResource] (main) CPUs: 1 singlethread: true maxMemory: 123M
2021-12-22 19:43:35,602 INFO  [io.quarkus] (main) cryostat-reports 1.0.0-SNAPSHOT on JVM (powered by Quarkus 2.5.3.Final) started in 3.889s. Listening on: http://0.0.0.0:8080
2021-12-22 19:43:35,605 INFO  [io.quarkus] (main) Profile prod activated. 
2021-12-22 19:43:35,606 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, vertx]
```

vs

```
$ MEMORY=1G sh run.sh 
+ set -e
+ '[' -z '' ']'
+ CPUS=1
+ '[' 1 -gt 1 ']'
+ GC=Serial
+ SINGLETHREAD_JFR_PARSE=true
+ '[' -z 1G ']'
+ podman run --user 0 --cpus 1 --memory 1G --publish 8080:8080 --env 'JAVA_OPTIONS=-XX:ActiveProcessorCount=1 -XX:+UseSerialGC -Dorg.openjdk.jmc.flightrecorder.parser.singlethreaded=true' --rm -it quay.io/cryostat/cryostat-reports:latest
exec java -XX:ActiveProcessorCount=1 -XX:+UseSerialGC -Dorg.openjdk.jmc.flightrecorder.parser.singlethreaded=true -cp . -jar /deployments/quarkus-run.jar
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/deployments/lib/main/org.jboss.slf4j.slf4j-jboss-logmanager-1.1.0.Final.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/deployments/lib/main/org.slf4j.slf4j-jdk14-1.7.30.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Slf4jLoggerFactory]
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2021-12-22 19:44:58,720 INFO  [io.cry.rep.ReportResource] (main) CPUs: 1 singlethread: true maxMemory: 247M
2021-12-22 19:44:59,397 INFO  [io.quarkus] (main) cryostat-reports 1.0.0-SNAPSHOT on JVM (powered by Quarkus 2.5.3.Final) started in 4.116s. Listening on: http://0.0.0.0:8080
2021-12-22 19:44:59,397 INFO  [io.quarkus] (main) Profile prod activated. 
2021-12-22 19:44:59,398 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, vertx]
```

it seems to be working as expected on my end - if I double the memory allocation then the application reports it has ~twice the `maxMemory`.

https://docs.oracle.com/javase/7/docs/api/java/lang/Runtime.html#maxMemory()

https://stackoverflow.com/questions/23701207/why-do-xmx-and-runtime-maxmemory-not-agree

hareetd replied:
> If the JVM isn't properly detecting its container memory limits then it would be choosing bad ratios for ex. heap size vs other memory pools, which could lead to weird OOM and performance issues when we're running workloads like this.
>
Hmm, that's weird. I'll need to poke around my machine and see why it's defaulting to ~8GB, thanks for the links.

hareetd replied:
It took me a while but I realized that I was running `podman` using `cgroup v1` which doesn't allow configuring resource limitation for rootless containers. I enabled `cgroup v2`, which also required updating my `Docker` setup (`podman` was good as is) and now the `--memory "${MEMORY}"` flag is working.

andrewazores replied:
> It took me a while but I realized that I was running `podman` using `cgroups v1` which doesn't allow configuring resource limitation for rootless containers. I enabled `cgroups v2`, which also required updating my `Docker` setup (`podman` was good as is) and now the `--memory "${MEMORY}"` flag is working.

Good find. If you have any links to documentation you found discussing this then maybe it would be worth it for us to link those in the main Cryostat README file in the "run in local podman" section or something.

hareetd commented:
That makes sense, I like the idea of applying a rejection heuristic based on recording size. 

I've been playing around with `run.sh` and with 8 cores and 8GB memory allocated, the 10MB, 150MB, and 480MB recordings take about 6.7s, 7.7s, and 87s to process report generation, respectively. I edited `smoketest.sh` to have the same allocation configuration but the previous failures are still observed (only dropdown report generation (and not View Report) works for the 10MB and 150MB recordings, with the 480MB failing in both cases). 

Do you have an idea as to why the reports container can smoothly process all three recordings over a simple HTTP setup but for some reason when clicking View Report with the full Cryostat setup it doesn't work? I guess that's more of an issue with the Cryostat backend/frontend since this container does its job well in isolation.

andrewazores commented:
> That makes sense, I like the idea of applying a rejection heuristic based on recording size.
> 
> I've been playing around with `run.sh` and with 8 cores and 8GB memory allocated, the 10MB, 150MB, and 480MB recordings take about 6.7s, 7.7s, and 87s to process report generation, respectively. I edited `smoketest.sh` to have the same allocation configuration but the previous failures are still observed (only dropdown report generation (and not View Report) works for the 10MB and 150MB recordings, with the 480MB failing in both cases).
> 
> Do you have an idea as to why the reports container can smoothly process all three recordings over a simple HTTP setup but for some reason when clicking View Report with the full Cryostat setup it doesn't work? I guess that's more of an issue with the Cryostat backend/frontend since this container does its job well in isolation.

That sounds to me like you're running into the HTTP timeout. Cryostat will only wait for up to 29 seconds to get a response from the report sidecar. Any longer and it cancels the request and reports a failure status back to the requesting client.

https://github.com/cryostatio/cryostat/pull/779/files#diff-08d5706f3bed9cd36f901546d703ac1e742d2fb33ff8330315e8e4483809ee25R73

The difference in behaviour for the dropdown vs View Report is still a bit strange sounding and I'll have to take a deeper look into that again, but that's going to be on the main Cryostat backend side of things.

andrewazores commented:
Here, roughly, is the entrypoint into where JMC displays that little range selection window after it determines that the given recording is too large.

https://github.com/openjdk/jmc/blob/1c6ccc9b644f7953ebb5f513aab5ada56760283b/application/org.openjdk.jmc.flightrecorder.ui/src/main/java/org/openjdk/jmc/flightrecorder/ui/RecordingLoader.java#L298

andrewazores commented:
I've added some logic to apply a configurable memory factor heuristic so that the service can bail out early if it thinks the provided recording is too large to process, so for that reason I have also updated this PR to close #6 - it's still possible to OOM, especially if the memory factor configuration is set to a low number, but I think this is about the best we can do.

I have also added some logic to check if the uploaded file was compresed, and if so, to first decompress it to disk before trying to process it. This should keep compressed recordings from exerting additional memory pressure, since the decompression doesn't also need to be done streaming in-memory while also processing the contents of the recording.

Finally, the elapsed time taken to decompress the file (if any) is checked, and if this takes too long then a `504` response is sent. If it didn't take too long then we continue to generate the report but now in a ForkJoinPool thread, and we only wait up to `timeout - elapsed` (where `elapsed` is essentially any decompression time) for that generation to complete. If it does not complete in time then we also respond with a `504`. I don't know if the underlying JMC report generation handles thread interruption well and will properly cancel processing here but at least we can respond nicely to clients.

andrewazores commented:
To create a compressed recording I simply did ex. `pigz --best --keep foo.jfr`, which will create `foo.jfr.gz`. Then `http -f :8080/report file@foo.jfr.gz` can be used as usual. I will eventually update the main Cryostat backend PR, or maybe open a new PR, to add configuration to enable compression on that end. I suspect that in a lot of cases, the extra time taken to compress the file on Cryostat's side and decompress it on the -reports side will probably be worth the reduced network transfer time. This also ties in neatly with https://github.com/cryostatio/cryostat/issues/732 .

jan-law commented on run.sh:
```diff
@@ -0,0 +1,35 @@
+fi
+
+if [ -z "${MEMORY}" ]; then
+    MEMORY="512M"
```
Since there's a tradeoff between having a lightweight container and having the ability to process large recordings, how did you decide on 512M as a default value? 
I'm also wondering how large of a recording most people would likely upload, and if this default amount of memory is enough to support most uploads without timing out.

andrewazores replied:
> Since there's a tradeoff between having a lightweight container and having the ability to process large recordings, how did you decide on 512M as a default value?

Well, `run.sh` is only for testing purposes - it won't affect anything in an actual deployment. My plan is to open an -operator PR at some point that adds configuration options for deploying -reports sidecars, and that configuration will include the number of replicas and the CPU/memory limits per replica. This way, the end user either adds no configuration and the deployment defaults to using subprocess generation within the Cryostat container, or the user adds configuration and makes their own choices about how many resources they're willing to spend for this feature.

> I'm also wondering how large of a recording most people would likely upload, and if this default amount of memory is enough to support most uploads without timing out.

That's hard to say and it will probably vary a lot between users. If a user wants to handle very large automated reports but doesn't want to pay for the required resources within their cluster then they still have the option of downloading the recording to their workstation and using its full resources for JMC to do the rules analysis.

jan-law replied:
> My plan is to open an -operator PR at some point that adds configuration options for deploying -reports sidecars, and that configuration will include the number of replicas and the CPU/memory limits per replica.

Thanks, adding memory/cpu configurations to the sidecar makes a lot of sense

hareetd reviewed (approved):
![image](https://user-images.githubusercontent.com/41028544/148476022-fb6e863c-8e7b-4087-8653-282547829166.png)

Tested it out using 8GB of memory, looks like it's working well for both compressed and uncompressed files. I wasn't able to trigger a `504` response since the memory heuristic with the default `memoryFactor=10` is working well and quickly responding with a `413` for files that are too large to process.

andrewazores commented:
You should be able to generate a `504` response pretty easily by doing something like `MEMORY=100M sh run.sh`. This is enough for the microservice to start up but obviously not enough to generate a result with even a 10MB recording file with the default 10 memoryFactor.

hareetd commented:
I tried that configuration out with a 10 MB recording and unfortunately, it still isn't enough to trigger a `504`. With the memory factor set to 10, the memory heuristic is still triggered (i.e. a `413`). However, if I lower the memory factor to 4 it passes the heuristic but causes an OOM (as you mentioned earlier). But by setting the timeout to 50ms I can artificially trigger a `504` due to decompression taking "too long". 

It seems the heuristic is doing its job well but I wonder in what conditions a `504` would be triggered during the actual report generation, with a more reasonable timeout of 30 seconds?

andrewazores commented:
Oh sorry, I read/wrote `504` but I had `413` in my head. Yea, I did have a combination I tested at some point that produced a timeout response while the timeout was still set to the default 30s, but I don't remember what that combination of memory allocation and recording size was - and it wouldn't necessarily behave exactly the same on your machine either since we have different CPUs etc. leading to different rules processing rates.

Another way to cause a 504 to happen could be to allocate a large amount of memory but a very small amount of CPU time. I just tested that out by modifying `run.sh` to pass `--cpus ${CPUS}` to Podman, but hardcode `ActiveProcessorCount=1` to the JVM. This has the effect of telling the JVM to behave as if it has one full CPU core, but in reality Podman will limit it to only get some proportion of CPU time. Then I did `CPUS=0.25 MEMORY=8G sh run.sh`, and posted my 149MB test file.

I think in real practical terms this is when we would most likely see 504s manifesting - when the -reports deployment has essentially overprovisioned memory for the amount of CPU available. It's also possible they could occur when the available memory is just barely enough to process the requested recording, but the JVM gets stuck with very little free allocation room between rules evaluations and ends up spending significant time on GC cycles (GC thrashing). This might be possible for smaller specific memory size/recording size with a single CPU core (even a whole one), particularly if the JVM is running with the Serial GC.

andrewazores commented:
Anyway, that last minor unit test addition was the last, last-minute change I thought of for this PR. I won't make any further alterations unless something comes up again in review.

hareetd reviewed (approved):
Nice, that CPU configuration gave me a `504` during report generation. I checked over the code and it looks good to go.

jan-law reviewed (approved):
> Nice, that CPU configuration gave me a 504 during report generation.

Same here. The 413 is working well too.

Tests also look good.
